traffic_ops/traffic_ops_golang/api/shared_handlers.go

```diff
@@ -144,6 +144,65 @@ func ReadHandler(reader Reader) http.HandlerFunc {
 	}
 }
 
+// DeprecatedReadHandler creates a net/http.HandlerFunc for the passed Reader object, and adds a deprecation
+// notice, optionally with a passed alternative route suggestion.
+func DeprecatedReadHandler(reader Reader, alternative *string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var alerts tc.Alerts
+		if alternative != nil {
+			alerts = tc.CreateAlerts(tc.WarnLevel, fmt.Sprintf("This endpoint is deprecated, please use %s instead", *alternative))
+		} else {
+			alerts = tc.CreateAlerts(tc.WarnLevel, "This endpoint is deprecated, and will be removed in the future")
+		}
+
+		inf, userErr, sysErr, errCode := NewInfo(r, nil, nil)
+		if userErr != nil || sysErr != nil {
+			if sysErr != nil {
+				log.Errorf(r.RemoteAddr + " " + sysErr.Error())
+			}
+			if userErr == nil {
+				userErr = errors.New(http.StatusText(errCode))
+			}
+			alerts.AddAlerts(tc.CreateErrorAlerts(userErr))
+			WriteAlerts(w, r, errCode, alerts)
+			log.Debugln(userErr.Error())
+			*r = *r.WithContext(context.WithValue(r.Context(), tc.StatusKey, errCode))
+			return
+		}
+
+		interfacePtr := reflect.ValueOf(reader)
+		if interfacePtr.Kind() != reflect.Ptr {
+			log.Errorf(r.RemoteAddr + " reflect: can only indirect from a pointer")
+			errCode = http.StatusInternalServerError
+			alerts.AddAlerts(tc.CreateErrorAlerts(errors.New(http.StatusText(errCode))))
+			WriteAlerts(w, r, errCode, alerts)
+			log.Debugln(userErr.Error())
+			*r = *r.WithContext(context.WithValue(r.Context(), tc.StatusKey, errCode))
+			return
+		}
+
+		objectType := reflect.Indirect(interfacePtr).Type()
+		obj := reflect.New(objectType).Interface().(Reader)
+		obj.SetInfo(inf)
+
+		results, userErr, sysErr, errCode := obj.Read()
+		if userErr != nil || sysErr != nil {
+			if sysErr != nil {
+				log.Errorf(r.RemoteAddr + " " + sysErr.Error())
+			}
+			if userErr == nil {
+				userErr = errors.New(http.StatusText(errCode))
+			}
+			alerts.AddAlerts(tc.CreateErrorAlerts(userErr))
+			WriteAlerts(w, r, errCode, alerts)
+			log.Debugln(userErr.Error())
+			*r = *r.WithContext(context.WithValue(r.Context(), tc.StatusKey, errCode))
+			return
```

Comment:
this is the same code as above lines 160-170 from what I can tell, can we collapse that into another utility func

Reply:
It's also very similar to `handleSimpleErr` So I think the best thing would be to abstract it out of there as well.